### PR TITLE
Add option to disallow overloading of terms and sorts

### PR DESCRIPTION
**Pull request by ajreynol**
This adds an option to disallow sorts and terms that have the same name. It is disabled by default, and thus the behavior of our parser is preserved.

This option make cvc5 *more* pedantic than the SMT-LIB 2.X standard. Future revisions (e.g. SMT-LIB 3.0) may disallow this, in which case this option can be enabled.

The motivation is that any proofs we generate will fail to check e.g. with ethos when this is the case.

There is one benchmark in SMT-LIB with this issue.

### CODE REVIEW

File: src/parser/symbol_table.cpp [hunk 1]
```diff
@@ -621,6 +637,12 @@ bool SymbolTable::Implementation::bindWithOverloading(const string& name,
     // the symbol manager.
     if (prev_bound_obj != obj)
     {
+      // If the previous overloaded symbol was a sort, we fail unconditionally
+      // in this case.
+      if (d_dummySortTerms.find(prev_bound_obj)!=d_dummySortTerms.end())
```
Comment (HanielB) [inline src/parser/symbol_table.cpp:642]: format

File: src/parser/sym_manager.cpp [hunk 2]
```diff
@@ -332,9 +333,21 @@ bool SymManager::bind(const std::string& name, cvc5::Term obj, bool doOverload)
   return d_implementation->getSymbolTable().bind(name, obj, doOverload);
 }
 
-void SymManager::bindType(const std::string& name, cvc5::Sort t)
+bool SymManager::bindType(const std::string& name, cvc5::Sort t, bool isUser)
 {
-  return d_implementation->getSymbolTable().bindType(name, t);
+  if (isUser && !d_termSortOverload)
+  {
+    // if a user sort and d_termSortOverload is false, we bind a dummy symbol
```
Comment (HanielB) [inline src/parser/sym_manager.cpp:340]: it's the term that is dummy, not the symbol?

File: src/parser/symbol_table.cpp [hunk 3]
```diff
@@ -621,6 +637,12 @@ bool SymbolTable::Implementation::bindWithOverloading(const string& name,
     // the symbol manager.
     if (prev_bound_obj != obj)
     {
+      // If the previous overloaded symbol was a sort, we fail unconditionally
+      // in this case.
```
Comment (HanielB) [inline src/parser/symbol_table.cpp:641]: It's not very clear what is happening here, because it is not very clear what d_dummySortTerms is for.

File: src/parser/symbol_table.h [hunk 4]
```diff
@@ -73,6 +73,14 @@ class CVC5_EXPORT SymbolTable
    */
   bool bind(const std::string& name, cvc5::Term obj, bool doOverload = false);
 
+  /**
```
Comment (HanielB) [inline src/parser/symbol_table.h:76]: document what this is doing and why it must exist

File: test/regress/cli/regress0/datatypes/pf-v2l60078.smt2 [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ; EXPECT: unsat
 (set-logic QF_DT)
-(declare-datatypes ((n 0) (l 0) (t 0)) (((z)) ((ln) (n (r l))) ((f))))
+(declare-datatypes ((n 0) (l 0) (t 0)) (((z)) ((ln) (nc (r l))) ((f))))
```
Comment (HanielB) [inline test/regress/cli/regress0/datatypes/pf-v2l60078.smt2:3]: I assume you are doing this to allow proofs to work with Ethos? Wouldn't it make more sense to disable the tester?

**Review by HanielB — CHANGES_REQUESTED** [review]

File: test/regress/cli/regress0/datatypes/pf-v2l60078.smt2 (hunk 5 above)
Comment (ajreynol) [reply test/regress/cli/regress0/datatypes/pf-v2l60078.smt2:3]: We had agreed that the "sensible" behavior is to disallow term/sort overloading, while we are still supporting it to not bother users.
In general our regressions should largely be the sensible use case unless we are specifically testing this corner case behavior.

Comment (HanielB) [reply test/regress/cli/regress0/datatypes/pf-v2l60078.smt2:3]: That sounds good. But if there isn't any `sat` regression with this kind of overloading, we should keep at least an `unsat` one to exercise that part of the parser.

Comment (ajreynol) [reply test/regress/cli/regress0/datatypes/pf-v2l60078.smt2:3]: Will add.

**Review by HanielB — APPROVED** [review]
